SwiftCompilerSources: support for ForwardingInstructions, ConversionInstructions, OwnershipTransitionInstruction

atrick opened the pull request:
No functional change intended. But this does fix holes in the existing utilities and make the Swift/C++ code base more consistent. That might improve the power of some analyses which could affect SIL patterns.

Namely, AccessUtils/MemAccessUtils now finds the actual reference root consistently.

atrick committed:
Declare local cl::opt flags `static`

atrick committed:
SwiftCompilerSources: Add enum BeginAccess.AccessKind

So we can switch over it.

atrick committed:
AddressUtils: refactor the findSingleInitializer API.

atrick committed:
Fix AccessUtils.swift to use the project_box reference root.

atrick committed:
Fix LetPropertyLowering to correctly find mark_uninitialized uses.

atrick commented:
@swift-ci test

meg-gupta commented on SwiftCompilerSources/Sources/SIL/ForwardingInstruction.swift:
```diff
@@ -397,3 +397,33 @@ extension LinearFunctionExtractInst: ForwardingInstruction {
+/// An instruction that transfers lifetime dependence from a single operand to a single result. The operand and result
+/// have the same identity, but they are not part of the same forwarded lifetime:
+/// copy_value, move_value, begin_borrow.
+public protocol OwnershipTransitionInstruction: UnaryInstruction {
```
nit: These are in "ForwardindInstruction".swift. Is there any other more appropriate file to put them ?

atrick replied:
No strong opinion. But I think it's important to understand the relationship between ForwardingInstruction and OwnershipTransitionInstruction. It's easy to get them confused and an instruction is either one or the other but never both

meg-gupta commented on lib/SIL/Utils/SILBridging.cpp:
```diff
@@ -402,11 +402,6 @@ static_assert((int)BridgedMemoryBehavior::MayWrite == (int)swift::MemoryBehavior
 static_assert((int)BridgedMemoryBehavior::MayReadWrite == (int)swift::MemoryBehavior::MayReadWrite);
 static_assert((int)BridgedMemoryBehavior::MayHaveSideEffects == (int)swift::MemoryBehavior::MayHaveSideEffects);
 
-static_assert((int)BridgedInstruction::AccessKind::Init == (int)swift::SILAccessKind::Init);
```
nit: can we keep the static_assert that c++ enum values are same as swift ones ?

atrick replied:
> nit: can we keep the static_assert that c++ enum values are same as swift ones ?

That would be great. Can you tell me how to do it?

meg-gupta replied:
I think it should work if you keep the original `AccessKind` in `SILBridging.h` and then create the swift enum this way: 

```
public var accessKind: AccessKind {
   AccessKind(rawValue: bridged.BeginAccessInst_getAccessKind().rawValue)!
 }
```

meg-gupta replied:
Huh, I guess that's not sufficient either to truly ensure these enum values  will be equivalent.

meg-gupta replied:
And way too complex and costly. So please ignore my comment!

eeckstein replied:
Maybe add a comment for `SILAccessKind` and `BeginAccessInst.AccessKind` that both must be kept in sync

meg-gupta reviewed (approved):
LGTM

atrick commented:
The deserializer crashed. Something to do with the swift version. Trying again.

atrick commented:
@swift-ci test

eeckstein commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -61,18 +61,22 @@ static llvm::cl::opt<bool> SILViewSILGenCFG(
-    "enable-deinit-devirtualizer", llvm::cl::init(false),
-    llvm::cl::desc("Enable the DeinitDevirtualizer pass."));
+static llvm::cl::opt<bool>
+    EnableDestroyHoisting("enable-destroy-hoisting", llvm::cl::init(false),
```
This looks like a wrong rebase: the EnableDestroyHoisting option was recently removed by Nate

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LetPropertyLowering.swift:
```diff
@@ -202,10 +202,28 @@ private extension RefElementAddrInst {
 }
 
 private extension Value {
+  func isReferenceDerived(from root: Value) -> Bool {
```
I think `referenceIsDerived(from:)` would be a better name

eeckstein reviewed (commented):
lgtm

atrick commented:
@swift-ci smoke test

atrick commented:
@swift-ci clean smoke test

eeckstein commented:
I'll take a look

atrick commented:
> I'll take a look

The bootstrapping1 compiler is miscompiling SIL.o.
The miscompiled function is FunctionEffects.description. This works around the problem:
```
+++ b/SwiftCompilerSources/Sources/SIL/Effects.swift
@@ -31,6 +31,7 @@ public struct FunctionEffects : CustomStringConvertible, NoReflectionChildren {
     self.sideEffects = nil
   }

+  @_optimize(none)
   public var description: String {
     var numArgEffects = escapeEffects.arguments.reduce(0, { max($0, $1.argumentIndex) }) + 1
     if let sideEffects = sideEffects {
```

atrick commented:
The call to `sideEffects.arguments.count` is being miscompiled.

I narrowed the problem down to a couple lines of code in a separate function that should be easy to compare SIL output and bisect:

```
+++ b/SwiftCompilerSources/Sources/SIL/Effects.swift
@@ -31,7 +31,11 @@ public struct FunctionEffects : CustomStringConvertible, NoReflectionChildren {
     self.sideEffects = nil
   }

+  @_optimize(none)
   public var description: String {
+    // print("REAL SIDE EFFECT COUNT \(sideEffects == nil ? 0 : sideEffects!.arguments.count)")
+    return getArgEffectsDescription() ? "GOOD EFFECT\n" : "BAD EFFECT\n"
+    /*
     var numArgEffects = escapeEffects.arguments.reduce(0, { max($0, $1.argumentIndex) }) + 1
     if let sideEffects = sideEffects {
       numArgEffects = max(numArgEffects, sideEffects.arguments.count)
@@ -56,6 +60,20 @@ public struct FunctionEffects : CustomStringConvertible, NoReflectionChildren {
       result += "[global: \(sideEffects.global)]\n"
     }
     return result
+     */
+  }
+
+  //!!!
+  @_silgen_name("__effects_decription")
+  @inline(never)
+  func getArgEffectsDescription() -> Bool {
+    var numArgEffects = escapeEffects.arguments.reduce(0, { max($0, $1.argumentIndex) }) + 1
+    //!!! After running the reduce loop above, sideEffects!.arguments.count is now magically 0!
+    // print("SIDEEFFECTCOUNT \(sideEffects == nil ? -1 : sideEffects!.arguments.count)")
+    if let sideEffects = sideEffects {
+      return sideEffects.arguments.count > 0
+    }
+    return false
   }
 }
```

atrick committed:
Code formatting cleanup related to ForwardingInstruction.

atrick committed:
Make MoveOnlyWrapperToCopyableBoxInst a ConversionInstruction.

This appears to be an obvious oversight.

atrick committed:
Add several instructions to ConversionOperation.

    MarkUninitializedInst
    MarkUnresolvedNonCopyableValueInst
    CopyableToMoveOnlyWrapperValueInst
    MoveOnlyWrapperToCopyableValueInst
    MoveOnlyWrapperToCopyableBoxInst

atrick committed:
Code formatting cleanup related to CopyingInstruction

atrick committed:
Add OwnershipTransitionInstruction.

These instructions carry lifetime dependence from a single operand to a single result. They are not forwarding
instructions because we use them to indicate the boundaries of a forwarded lifetime.

atrick committed:
Fix LifetimeDependenceUtils to use OwnershipTransitionInstruction

atrick committed:
Fix AccessUtils Value.referenceRoot to handle ForwardingInstructions.

This fixes all analyses that use AccessUtils in the presence of new
forwarding instructions. This is also needed for consistency with the
C++ source base. And for general sanity.

Utilities that walk the use-def chain should never hard-code a list of
opcodes that happened to work with some specific pass. Passes should
never assume that a basic SIL utility handles specific SIL operations
a certain way. Instead, the utility needs to be defined in terms of
SIL semantics. In this case, the utility is supposed to handle all
instructions that semantically forward ownership of a reference.

atrick commented:
@swift-ci test

atrick commented:
Failed Tests (1):
  Swift-validation(linux-x86_64) :: stdlib/RangeSet.swift

# *** Signal 6: Backtracing from 0x7f6a88ffc00b... done ***

atrick commented:
@swift-ci test linux platform

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test

atrick committed:
Fix MarkUnresolvedReferenceBindingInst handling

It should be treated as a marker instruction. This looks like a mistake.

atrick committed:
Add ForwardingInstruction.preservesIdentity.

Mirror isIdentityPreservingRefCase.

I don't know why this does not apply to boxes and existentials, but am convervatively staying consistent with the
current behavior.

atrick commented:
@swift-ci test

atrick reviewed (commented):
/cc @gottesmm 
[Fix MarkUnresolvedReferenceBindingInst handling](https://github.com/apple/swift/pull/71826/commits/a048455ad4d77fc484bbf723053d08f0360ff38f)
